File: day06/main.py
#! /usr/bin/env python3
import re
import itertools


def day06a(groups):
    total = 0
    for g in groups:
        s = set()
        for c in g:
            if c in "abcdefghijklmnopqrstuvwxyz":
                s.add(c)
        total += len(s)
    return total


def day06b(groups):
    total = 0
    for g in groups:
        intersectionSet = None
        for person in g.splitlines(keepends=False):
            s = set()
            for c in person:
                s.add(c)
            if intersectionSet is None:
                intersectionSet = s
            else:
                intersectionSet = intersectionSet & s
        total += len(intersectionSet)
    return total


if __name__ == "__main__":
    f = open("./input.txt")
    data = f.read()
    groups = re.split("\n[ \t]*\n", data)
    print(f"{day06a(groups)} questions were answered in part A")
    print(
        f"{day06b(groups)} questions were answered by all members of a group in part B"
    )

File: day06/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"regexp"
	"strings"

	"github.com/kentquirk/stringset/v2"
)

func day06a(forms []string) int {
	total := 0
	for _, group := range forms {
		set := stringset.New()
		for _, c := range group {
			if c >= 'a' && c <= 'z' {
				set.Add(string(c))
			}
		}
		total += set.Length()
	}
	return total
}

func day06b(forms []string) int {
	total := 0
	for _, group := range forms {
		people := strings.Split(group, "\n")
		var intersectionSet *stringset.StringSet
		for _, person := range people {
			set := stringset.New()
			for _, c := range person {
				if c >= 'a' && c <= 'z' {
					set.Add(string(c))
				}
			}
			if intersectionSet == nil {
				intersectionSet = set
			} else {
				intersectionSet = intersectionSet.Intersection(set)
			}
		}

		total += intersectionSet.Length()
	}
	return total
}

func main() {
	f, err := os.Open("./input.txt")
	if err != nil {
		log.Fatal(err)
	}
	b, err := ioutil.ReadAll(f)
	if err != nil {
		log.Fatal(err)
	}
	splitpat := regexp.MustCompile("\n[[:blank:]]*\n")
	forms := splitpat.Split(string(b), -1)
	fmt.Printf("%d questions were answered in part A\n", day06a(forms))
	fmt.Printf("%d questions were answered the same by all group participants in part B\n", day06b(forms))
}
